tests/test_updater_top_level_update.py

```diff
@@ -424,6 +424,47 @@ def test_new_snapshot_version_rollback(self) -> None:
 
         self._assert_version_equals(Snapshot.type, 2)
 
+    def test_new_snapshot_fast_foward_recovery(self) -> None:
+        """Test snapshot fast-forward recovery using key rotation.
+
+        The snapshot recovery requires the snapshot and timestamp key rotation.
+        It is made by the following steps:
+        - Remove the snapshot and timestamp keys
+        - Create and add a new key for snapshot and timestamp
+        - Rollback snapshot version
+        - Bump and publish root
+        - Bump the timestamp
+        """
+
+        # attacker updates to a higher version (bumping timestamp is required)
+        self.sim.snapshot.version = 99999
+        self.sim.update_timestamp()
+
+        # client refreshes the metadata and see the new snapshot version
+        self._run_refresh()
+        self._assert_version_equals(Snapshot.type, 99999)
+
+        # repo add new snapshot and timestamp keys and recovers snapshot version
+        self.sim.root.roles["snapshot"].keyids.clear()
+        self.sim.signers["snapshot"].clear()
+        self.sim.root.roles["timestamp"].keyids.clear()
+        self.sim.signers["timestamp"].clear()
+        snapshot_key, snapshot_signer = self.sim.create_key()
+        self.sim.root.add_key("snapshot", snapshot_key)
+        self.sim.add_signer("snapshot", snapshot_signer)
+        timestamp_key, timestamp_signer = self.sim.create_key()
+        self.sim.root.add_key("timestamp", timestamp_key)
+        self.sim.add_signer("timestamp", timestamp_signer)
+        self.sim.snapshot.version = 1
+        self.sim.root.version += 1
+        self.sim.publish_root()
+
+        self.sim.update_timestamp()
```

Comment:
This looks correct but for clarity following might be better (root update just contains the key changes: then we rollback snapshot version, signing with the new keys)

```suggestion
        self.sim.root.version += 1
        self.sim.publish_root()

        self.sim.snapshot.version = 1
        self.sim.update_timestamp()
```